### PR TITLE
Add initial support for dynamicly loading comments

### PR DESCRIPTION
**Pull request by CarlosM10011**
This pull request adds initial support for rendering comments dynamically via a java servlet. Currently it only has a few comments hard coded for demonstration purposes.

### CODE REVIEW

File: portfolio/src/main/webapp/script.js [hunk 1]
```diff
@@ -26,3 +26,37 @@ function addRandomGreeting() {
...
+    if (response.status < 200 ||
+        response.status > 299) {  // HTTP error codes in the 200 range are okay.
+      throw response.status;
+    }
+    const rawComments = await response.json();
+    for (let i = 0; i < rawComments.length; i++) {
+      const newComment = commentTemplate.content.cloneNode(true);
+      const subjectElement = newComment.getElementById('subject');
+      const bodyElement = newComment.getElementById('body');
+      subjectElement.textContent =
```
Comment (hzimz) [inline portfolio/src/main/webapp/script.js:49]: Prefer templates over string concatenation: https://google.github.io/styleguide/jsguide.html#features-strings-template-strings

File: portfolio/src/main/webapp/script.js [hunk 2]
```diff
@@ -26,3 +26,37 @@ function addRandomGreeting() {
...
+    const response = await fetch('/data');
+    if (response.status < 200 ||
+        response.status > 299) {  // HTTP error codes in the 200 range are okay.
+      throw response.status;
+    }
+    const rawComments = await response.json();
+    for (let i = 0; i < rawComments.length; i++) {
+      const newComment = commentTemplate.content.cloneNode(true);
+      const subjectElement = newComment.getElementById('subject');
+      const bodyElement = newComment.getElementById('body');
```
Comment (hzimz) [inline portfolio/src/main/webapp/script.js]: Move this to just above line 47 and generally, delay declaration of variables for as long as possible (preferably until right above the line they're needed)

File: portfolio/src/main/webapp/script.js [hunk 3]
```diff
@@ -26,3 +26,37 @@ function addRandomGreeting() {
...
+      subjectElement.textContent =
+          rawComments[i].name + ' on ' + rawComments[i].created;
+      bodyElement.textContent = rawComments[i].message;
+      commentsContainer.appendChild(newComment);
+    }
+  } catch (e) {
+    console.error('Error:');
```
Comment (hzimz) [inline portfolio/src/main/webapp/script.js]: Any reason these two are separate instead of one console.error call?

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 4]
```diff
@@ -14,19 +14,33 @@
...
 @WebServlet("/data")
 public class DataServlet extends HttpServlet {
 
   @Override
   public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
-    response.setContentType("text/html;");
-    response.getWriter().println("<h1>Hello world!</h1>");
+    // Hard coded comment data for now.
+    Comment[] comments = new Comment[3];
+    for (int i = 0; i < 3; i++) {
```
Comment (hzimz) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: Any reason why you want to use this logic instead of just initializing the array to have the three comments?

**Review by hzimz — CHANGES_REQUESTED** [review]

File: portfolio/src/main/webapp/script.js [hunk 5]
```diff
@@ -26,3 +26,37 @@ function addRandomGreeting() {
...
+    commentsContainer.innerText =
+        'An error occured loading comments. Please try again later.';
+    return;
+  }
+}
+
+                /** At the monent, this only loads comments */
+                window.onload = () => {
+  addComments();
+}
```
Comment (tinagrasso) [inline portfolio/src/main/webapp/script.js]: whitespace.

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 6]
```diff
@@ -30,14 +30,12 @@
...
     // Hard coded comment data for now.
-    Comment[] comments = new Comment[3];
-    for (int i = 0; i < 3; i++) {
-      comments[i] =
-          new Comment(
-              "Carlos" + Integer.toString(i + 1),
-              new Date(),
-              "This is comment number " + Integer.toString(i + 1));
-    }
+    Comment[] comments =
```
Comment (tinagrasso) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: Use an ArrayList here and make it immutable.

When choosing a container type, always prefer a Collections subclass. Object arrays are a rarely needed micro-optimization and inferior to other container types. For cases like this where order and duplicates don't matter, use an ArrayList.

As a side note, since I've seen a lot of students do this before, when choosing between ArrayList and LinkedList, always choose ArrayList. It's very rare for a LinkedList to outperform an ArrayList.

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 7]
```diff
@@ -30,14 +30,12 @@
...
+        new Comment[] {
+          new Comment("Fred", new Date(), "This is comment number 1"),
+          new Comment("George", new Date(), "This is comment number 2"),
+          new Comment("Ron", new Date(), "This is comment number 3")
+        };
     Gson gson = new Gson();
     String output = gson.toJson(comments);
     response.setContentType("text/json;");
```
Comment (tinagrasso) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: Make the content type a constant.

```
private final String CONTENT_TYPE = "text/json";
```

File: portfolio/pom.xml [hunk 8]
```diff
@@ -17,6 +17,11 @@
   </properties>
 
   <dependencies>
+	  <dependency>
+		    <groupId>com.google.code.gson</groupId>
+		      <artifactId>gson</artifactId>
+		        <version>2.8.6</version>
+		</dependency>
```
Comment (tinagrasso) [inline portfolio/pom.xml]: fix indents

File: portfolio/src/main/java/com/google/sps/data/Comment.java [hunk 9]
```diff
@@ -0,0 +1,43 @@
...
+
+  String getName() {
+    return this.name;
+  }
+
+  String getMessage() {
+    return this.message;
+  }
+
+  Date getCreated() {
```
Comment (tinagrasso) [inline portfolio/src/main/java/com/google/sps/data/Comment.java]: Add visibility modifiers.

File: portfolio/src/main/webapp/index.html [hunk 10]
```diff
@@ -1,18 +1,28 @@
...
-      <p>This is my portfolio.</p>
-      <p>Click here to get a random message:</p>
-      <button onclick="addRandomGreeting()">Hello</button>
-      <div id="greeting-container"></div>
+        <h1>My Portfolio</h1>
+        <p>This is my portfolio.</p>
+        <p>Click here to get a random message:</p>
+        <button onclick="addRandomGreeting()">Hello</button>
+        <div id="fact-container"></div>
     </div>
```
Comment (tinagrasso) [inline portfolio/src/main/webapp/index.html]: Changes in this file that don't have anything to do with the feature it addresses should be put in a different PR.

File: portfolio/src/main/webapp/index.html [hunk 11]
```diff
@@ -1,18 +1,28 @@
...
-  </body>
+    <section id="comments-container">
+        <h1>Comments</h1>
+        <template id="comment-template">
```
Comment (tinagrasso) [inline portfolio/src/main/webapp/index.html:28]: Very cool! I didn't know this tag existed.

File: portfolio/src/main/webapp/index.html [hunk 12]
```diff
@@ -1,18 +1,28 @@
 <!DOCTYPE html>
 <html>
-  <head>
+
+<head>
```
Comment (tinagrasso) [inline portfolio/src/main/webapp/index.html]: Fix the indents for this file.

File: portfolio/src/main/webapp/script.js [hunk 13]
```diff
@@ -26,3 +26,36 @@ function addRandomGreeting() {
...
 }
+
+const addComments =
+    async () => {
+  try {
+    const commentTemplate = document.getElementById('comment-template');
+    const commentsContainer = document.getElementById('comments-container');
+    const response = await fetch('/data');
+    if (response.status < 200 ||
+        response.status > 299) {  // HTTP error codes in the 200 range are okay.
```
Comment (tinagrasso) [inline portfolio/src/main/webapp/script.js]: This can be improved.

```
if (!response.ok()) {
  // throw the error.
}
```

In general, when you throw errors, you should make them as descriptive as possible so that they're easier to debug. In this case, you're only throwing a status code, but no accompanying error message. The best thing to do here would be to encapsulate the error in a new object with a message about what broke, the status code and error message.

```
throw new Error("Error occurred while fetching comment data.", response.status, response.statusText);
```

File: portfolio/src/main/webapp/script.js [hunk 14]
```diff
@@ -26,3 +26,36 @@ function addRandomGreeting() {
...
+    console.error(e);
+    commentsContainer.innerText =
+        'An error occured loading comments. Please try again later.';
+    return;
+  }
+}
+
+                /** At the monent, this only loads comments */
+                window.onload = () => {
+  addComments();
```
Comment (tinagrasso) [inline portfolio/src/main/webapp/script.js]: Fix indentation.

File: portfolio/src/main/webapp/script.js [hunk 15]
```diff
@@ -26,3 +26,36 @@ function addRandomGreeting() {
...
+    }
+  } catch (e) {
+    console.error(e);
+    commentsContainer.innerText =
+        'An error occured loading comments. Please try again later.';
+    return;
+  }
+}
+
+                /** At the monent, this only loads comments */
```
Comment (tinagrasso) [inline portfolio/src/main/webapp/script.js]: Add a period at the end of comments. They should be sentence like.

**Review by tinagrasso — CHANGES_REQUESTED** [review]

File: portfolio/src/main/webapp/index.html (hunk 10 above)
Comment (CarlosM10011) [reply portfolio/src/main/webapp/index.html]: Hmm, this is interesting. I think diff got confused because my auto indent broke. This pr should only be adding comment support in this file.

**Review by hzimz — APPROVED** [review]

**Review by tinagrasso — APPROVED** [review]
